### PR TITLE
implement estimation of relative overfitting for more preds-ish (#433)

### PR DESCRIPTION
**Pull request by larskotthoff**
It would be good if somebody could check this, especially wrt extracting train/test data -- not sure at all about this part.

### CODE REVIEW

**Comment by larskotthoff** [comment]
No idea why Travis fails -- seems like it's something completely unrelated and I can't reproduce it locally.

**Comment by jakob-r** [comment]
I think you spotted why it failed - because caret changed their defaults.

**Comment by larskotthoff** [comment]
@berndbischl and I will talk about this.

**Comment by ja-thomas** [comment]
Please describe what exactly is the purpose of this. And it needs a better title (Any issues connected to this?

**Comment by berndbischl** [comment]
@larskotthoff 
please answer the question here. this is soon getting extremely old. we already have conflicts. 
if we dont hurry up, this will simply be lost...

can you please 
a) explain what goes on here
b) resolve the conflicts

**Comment by larskotthoff** [comment]
This is a cleanup of the relative overfitting code to make it work better with the objects mlr provides. No issue related to this.

**Comment by mllg** [comment]
1. No need to document `NULL` here. This way you don't see the function definition in the docs.
2. The object is document as `obj`, but `predish` is used in the code.
3. You don't need `assertClass` on the first object in S3 methods.
4. See line comments.

File: R/relativeOverfitting.R [hunk 1]
```diff
@@ -19,37 +23,64 @@
...
-    df = (perf.test - perf.train) / (perf.permuted - perf.train)
-    names(df) = stri_paste("relative.overfit", mids, sep = ".")
-    cbind(data.frame(iter = i), df)
+  iterations = unique(predish$data$iter)
+  rbindlist(lapply(iterations, function(i) {
+    data = predish$data[predish$data$iter == i & predish$data$set == "test",]
+    pred.test = makePrediction(task$task.desc, row.names(data), data$id, data$truth, predish$predict.type, predish$predict.threshold, data$response, predish$time[i])
+
+    data = predish$data[predish$data$iter != i & predish$data$set == "test",]
+    pred.train = makePrediction(task$task.desc, row.names(data), data$id, data$truth, predish$predict.type, predish$predict.threshold, data$response, predish$time[i])
```
Comment (mllg) [inline R/relativeOverfitting.R]: Sure about `predish$time[i]`?

File: R/relativeOverfitting.R [hunk 2]
```diff
@@ -19,37 +23,64 @@
...
+  measures = checkMeasures(measures, task)
+  mids = vcapply(measures, function(m) m$id)
+
+  perf.test = performance(predish, measures = measures)
+  perf.train = performance(pred.train, measures = measures)
+
+  nrows = nrow(predish$data) + nrow(pred.train$data)
+  pred.permuted = predish
+  pred.permuted$data = data.frame(truth = rep(c(predish$data$truth, pred.train$data$truth), each = nrows),
+    response = rep(c(predish$data$response, pred.train$data$response), times = nrows))
```
Comment (mllg) [inline R/relativeOverfitting.R]: This seems odd. Are you sure you don't want to call `sample()` somewhere?

Comment (mllg) [inline R/relativeOverfitting.R]: (Without having a deeper look into the objects): I suspect `predish$data` and `pred.train$data` have the same number of rows -- why do you need to call `rep` once with `each` and once with `times`? Maybe a short comment would clarify this line.

File: R/relativeOverfitting.R (hunk 1 above)
Comment (larskotthoff) [inline R/relativeOverfitting.R]: Yes. Is that not the way you're supposed to access it?

File: R/relativeOverfitting.R (hunk 2 above)
Comment (larskotthoff) [inline R/relativeOverfitting.R]: It's doing all permutations, so no need to sample. I'll add a comment.

**Comment by larskotthoff** [comment]
Do we want this or not?

**Pull request merged**